File: src/chapters/minimization/content/code/simplex.jl
# Nelder-Mead downhill simplex minimization.
function nelder_mead(f, x0; step = 0.5, tol = 1e-10, max_iter = 400)
    n = length(x0)
    pts = [float.(x0)]
    for i in 1:n
        p = float.(x0); p[i] += step; push!(pts, p)
    end
    fv = [f(p) for p in pts]
    for _ in 1:max_iter
        ord = sortperm(fv)
        pts, fv = pts[ord], fv[ord]
        abs(fv[end] - fv[1]) < tol && break
        c = sum(pts[1:end-1]) / n                  # centroid of best n points
        xr = c + (c - pts[end]); fr = f(xr)        # reflect
        if fr < fv[1]
            xe = c + 2*(c - pts[end]); fe = f(xe)  # expand
            pts[end], fv[end] = fe < fr ? (xe, fe) : (xr, fr)
        elseif fr < fv[end-1]
            pts[end], fv[end] = xr, fr
        else
            xc = c + 0.5*(pts[end] - c); fc = f(xc)  # contract
            if fc < fv[end]
                pts[end], fv[end] = xc, fc
            else                                     # shrink toward best
                for i in 2:n+1
                    pts[i] = pts[1] + 0.5*(pts[i] - pts[1]); fv[i] = f(pts[i])
                end
            end
        end
    end
    return pts[1]
end
f = v -> (v[1] - 1)^2 + (v[2] - 2)^2
println(nelder_mead(f, [0.0, 0.0]))               # -> [1, 2]

File: src/chapters/minimization/content/code/simplex.rs
// Nelder-Mead downhill simplex minimization.
fn nelder_mead<F: Fn(&[f64]) -> f64>(
    f: F, x0: &[f64], step: f64, tol: f64, max_iter: usize,
) -> Vec<f64> {
    let n = x0.len();
    let mut pts: Vec<Vec<f64>> = vec![x0.to_vec()];
    for i in 0..n {
        let mut p = x0.to_vec(); p[i] += step; pts.push(p);
    }
    let mut fv: Vec<f64> = pts.iter().map(|p| f(p)).collect();
    for _ in 0..max_iter {
        let mut ord: Vec<usize> = (0..=n).collect();
        ord.sort_by(|&a, &b| fv[a].partial_cmp(&fv[b]).unwrap());
        pts = ord.iter().map(|&i| pts[i].clone()).collect();
        fv = ord.iter().map(|&i| fv[i]).collect();
        if (fv[n] - fv[0]).abs() < tol { break; }
        let mut c = vec![0.0; n];                       // centroid of best n points
        for p in pts.iter().take(n) { for j in 0..n { c[j] += p[j] / n as f64; } }
        let xr: Vec<f64> = (0..n).map(|j| c[j] + (c[j] - pts[n][j])).collect();  // reflect
        let fr = f(&xr);
        if fr < fv[0] {
            let xe: Vec<f64> = (0..n).map(|j| c[j] + 2.0 * (c[j] - pts[n][j])).collect();  // expand
            let fe = f(&xe);
            if fe < fr { pts[n] = xe; fv[n] = fe; } else { pts[n] = xr; fv[n] = fr; }
        } else if fr < fv[n - 1] {
            pts[n] = xr; fv[n] = fr;
        } else {
            let xc: Vec<f64> = (0..n).map(|j| c[j] + 0.5 * (pts[n][j] - c[j])).collect();  // contract
            let fc = f(&xc);
            if fc < fv[n] { pts[n] = xc; fv[n] = fc; }
            else {                                       // shrink toward best
                for i in 1..=n {
                    pts[i] = (0..n).map(|j| pts[0][j] + 0.5 * (pts[i][j] - pts[0][j])).collect();
                    fv[i] = f(&pts[i]);
                }
            }
        }
    }
    pts[0].clone()
}

fn main() {
    let f = |v: &[f64]| (v[0] - 1.0).powi(2) + (v[1] - 2.0).powi(2);
    println!("{:?}", nelder_mead(f, &[0.0, 0.0], 0.5, 1e-10, 400));   // -> [1, 2]
}
